resource-aggregate/cqrs/aggregate/aggregate.go

```diff
@@ -174,39 +169,65 @@ func (a *Aggregate) handleCommandWithAggrModel(ctx context.Context, cmd Command,
 		newVersion++
 	}
 
-	events = make([]eventstore.Event, 0, 32)
+	previousSnapshotEvent, ok := amodel.TakeSnapshot(newVersion)
 	if amodel.numEvents >= a.numEventsInSnapshot {
-		snapshotEvent, ok := amodel.TakeSnapshot(newVersion)
 		if ok {
-			concurrencyException, err := a.store.SaveSnapshot(ctx, a.groupID, a.aggregateID, snapshotEvent)
+			status, err := a.store.Save(ctx, previousSnapshotEvent)
 			if err != nil {
 				return nil, false, fmt.Errorf("cannot save snapshot: %w", err)
 			}
-			if concurrencyException {
+			switch status {
+			case eventstore.NeedSnapshot:
+				return nil, false, fmt.Errorf("cannot need snapshot during store a snapshot")
+			case eventstore.ConcurrencyException:
 				return nil, true, nil
 			}
 			newVersion++
-			events = append(events, snapshotEvent)
+			a.createdSnapshot = previousSnapshotEvent
 		}
 	}
 
 	newEvents, err := amodel.HandleCommand(ctx, cmd, newVersion)
 	if err != nil {
 		return nil, false, fmt.Errorf("cannot handle command by model: %w", err)
 	}
-
-	if len(newEvents) > 0 {
-		concurrencyException, err := a.store.Save(ctx, a.groupID, a.aggregateID, newEvents)
+	if newEvents == nil {
+		if a.createdSnapshot != nil && a.createdSnapshot.Version()+1 == newVersion {
+			events = append(events, a.createdSnapshot)
+			a.createdSnapshot = nil
+		}
+		return events, false, nil
+	}
+	status, err := a.store.Save(ctx, newEvents...)
+	if err != nil {
+		return nil, false, fmt.Errorf("cannot save events: %w", err)
+	}
+	switch status {
+	case eventstore.NeedSnapshot:
+		if previousSnapshotEvent == nil {
+			return nil, false, fmt.Errorf("cannot save events[%v]: snapshot is not supported", newEvents)
+		}
+		if a.createdSnapshot != nil && a.createdSnapshot.Version()+1 == newEvents[0].Version() {
+			return nil, false, fmt.Errorf("cannot save events[%v]: snapshot[%v] follows created previous snapshot[%v]", newEvents, previousSnapshotEvent, a.createdSnapshot)
```

Comment:
Do you really need to dump both snapshots and new events? How will it help you to investigate an error?
"snapshot follows previously created snapshot", but still, when this can happen?